File: June-LeetCoding-Challenge-2020/Week2/231-Power-of-Two.cpp
// Time Complexity: O(logn)
// Space Complexity: O(1)
// Runtime: 4 ms, faster than 37.98%
// Memory Usage: 5.9 MB, less than 59.80%
class Solution {
public:
    bool isPowerOfTwo(int n) {

        if(n == 1)
            return true;
        
        int comp = 1;
        
        while(comp <= n){
            if(comp == n)
                return true;
            if(comp > INT_MAX/2)
                return false;
            comp *= 2;
        }
        
        return false;
    }
};

File: June-LeetCoding-Challenge-2020/Week2/231-power-of-two.py
class Solution:
    # Sol 1. Tow's Complement
    def isPowerOfTwo(self, n: int) -> bool:
        if n is 0:
            return False
        return n & (-n) == n
    # TC: O(1)
    # SC: O(1)


    def isPowerOfTwo(self, n: int) -> bool:
        if n is 0:
            return False
       
        while n%2 == 0:
            n = n//2
        
        return n==1
    # TC: O(logn)
    # SC: O(1)